02_bubble/jonathanheaden/main.go

```diff
@@ -0,0 +1,31 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func bubble(s []int) []int {
+	done := false
+	if len(s) == 0 {
+		return s
+	}
+	for !done {
+		sw := false
```

Comment:
The additional variable might not necessary, you might be able to track sort completion with a single variable.

Reply:
I use the two vars for tracking overall completion between the two loops, done only gets set true when all iterations of the the `sw` swap flag in the inner loop return false (no more sorting required).

Reply:
I think you might be able to achieve the same thing with a single variable as @nicholascross mentioned.

```go
sorted := false
for !sorted {
    sorted = true
    for i := 0; i < len(r)-1; i++ {
        if r[i] > r[i+1] {
            sorted = false
            r[i], r[i+1] = r[i+1], r[i]
        }
    }
}
```

I haven't tested it out properly but I think that might do the job.